spmd/tensor/ops/tensor_ops.py

```diff
@@ -278,3 +279,176 @@ def prop_slice_scatter(op_schema: OpSchema) -> OutputSharding:
+            )
+            for s in result.schema_suggestions
+        ]
+    return result
+
+
+@register_prop_rule("aten.index.Tensor")
+def prop_index(op_schema: OpSchema) -> OutputSharding:
+    """
+    Expect replicated on the first input; _mostly_ pointwise on the second input.
+    TODO: exception: when the dtype of second input is "bool", then a torch.nonzero needs to be triggered first.
+    """
+    values_spec, multi_indices_spec = op_schema.args_schema
+    assert isinstance(values_spec, DTensorSpec)
+    multi_indices_spec = cast(
+        Union[DTensorSpec, List[Optional[DTensorSpec]]], multi_indices_spec
+    )
+
+    if isinstance(multi_indices_spec, list):
+        valid_indices_spec: List[Tuple[int, DTensorSpec]] = [
+            (i, a) for i, a in enumerate(multi_indices_spec) if a is not None
+        ]
+    else:
+        valid_indices_spec = [(0, multi_indices_spec)]
+        multi_indices_spec = [multi_indices_spec]
+
+    # TODO: not possible to check dtype for indices here, so assuming its not bool.
+
+    # we're good if
+    # 1. values[dim] is replicated for each dim
+    # 2. other dimensions of values is not sharded on a mesh dimension that indices is also sharded in.
+    # 3. no pending reduction on indices
+
+    # 1. All indices have to be sharded equally. Moreover, indices can be broadcast.
+    #    Here, we piggyback on the pointwise sharding rule for indices.
+    indices_out = pointwise_rule(
+        OpSchema(
+            args_schema=tuple(v[1] for v in valid_indices_spec),
+            kwargs_schema={},
+        )
+    )
+    need_reshard_on_indices = indices_out.output_spec is None
+
+    # N.B.: the weird typing below is needed to make pyre happy.
+    indices_suggestion: Union[
+        List[Optional[DTensorSpec]], List[DTensorSpec]
+    ] = multi_indices_spec
+    if not need_reshard_on_indices:
+        # this means that our inputs are already sharded properly and we will use that as our indices_spec
+        assert isinstance(indices_out.output_spec, DTensorSpec)
+        indices_spec: DTensorSpec = indices_out.output_spec
+    else:
+        assert indices_out.schema_suggestions is not None
+        valid_indices_suggestion = indices_out.schema_suggestions[0]
+        for i, v in enumerate(valid_indices_suggestion.args_spec):
+            indices_suggestion[valid_indices_spec[i][0]] = v
+        # we'll need to call pointwise_rule again to see what's our ideal indices_spec and then
+        # use that to compute our ideal values_spec
+        indices_output_spec = pointwise_rule(
+            valid_indices_suggestion
+        ).output_spec
+        assert isinstance(indices_output_spec, DTensorSpec)
+        indices_spec = indices_output_spec
+
+    lookup_dims = set(v[0] for v in valid_indices_spec)
+
+    need_reshard_on_values = tuple(
+        (
+            isinstance(vp, Shard)
+            and (vp.dim in lookup_dims or isinstance(ip, Shard))
```

Comment:
could you document a bit more on the semantics of this rule, so it seems like we only reshard values_spec to replicate in the case the value's sharding dim is the lookup dim in indices?

Reply:
Yes, let me add a more thorough comment here.